### ISSUE TITLE
Backing up user data

### ISSUE THREAD

**Issue by TiemenSch** [jupyterhub/zero-to-jupyterhub-k8s#2002]
### Proposed change

Some recommended way or form of backing up and restoring user data in a Kubernetes-volume-friendly manner to a certain point in time.

It would be nice if there was some recommended form of backing up and restoring user volumes to for instance an Amazon S3 bucket or other form of off-site storage.

Since users are added and removed from the Hub over time, it makes sense to support this feature from the Hub and not have to select each (dynamically created) volume in your cloud provider's settings manually to turn on periodic backups. 

My initial thoughts would be that it's best to leave the user-PVC-PV bookkeeping to the Hub, since that may become unwieldy for larger numbers of users very fast. Perhaps a Kubernetes Job to run these backups periodically and a similar one-time thing for restoring backups.

### Alternative options

Currently, backing up user data is either up to the user themselves or to the sysadmins running ztjh (or any JupyterHub deployment).

Some cloud providers have a periodic backup option on their volumes, but they may require you selecting each volume and turning on the backups manually, which is quite the hassle.

### Who would use this feature?

End-users and a form of peace of mind for sysadmins.

**Comment by TiemenSch** [comment]
Currently, the only "backup" mentioned in the docs is that you should backup your database before large upgrades. User storage backup is pretty much left for the admin to figure out. Which makes sense, since there is such a plethora of methods to do or achieve this depending on your storage solution.

However, I think that at least describing 'one way to do it' may be enough to help people get started.

**Comment by consideRatio** [comment]
I agree, storage backup and such is a topic that the guide could cover to some degree. While I think it's hard to establish a recommendation, I think we still can give some suggestions of what can be done.

**Comment by agnewp** [comment]
as a admin trying to figure out how to migrate jupyter hub from one cluster to another, i would very much appreciate some guidance on how to backup and restore user data without contacting each person and telling them i will delete everything and they had better save what they can now

**Comment by consideRatio** [comment]
If you use for example GitHub OAauthenticator, then I don't think you need to save the hub disk. It just lists the users that has logged in once, but since who has access is defined by the OAuth2 application and such it doesn't matter much what is said in the current hub database about past users.

What matters is probably the user storage. The user storage assuming default configuration will be many separate PVs created through request from PVCs. Depending on your cloud provider you can do different things. On GKE you could do "disk snapshots" of the GCP disks mapped byt he PVs and then recreate them in another cluster and such. I don't know of any automated way to do something...


Hmm... in modern k8s versions, there are also VolumeSnapshots, but I don't think that will be very helpful...

Hmm... I think my best suggestion would be to setup NFS storage external to the k8s cluster and load all user storage into their own folder in the NFS storage, and then use NFS storage in the new cluster. Perhaps...

It is quite messy to migrate a large set of PVCs, and I would say it is a bit out of scope to be directly addressed by the z2jh.jupyter.org guide.

**Comment by agnewp** [comment]
forget the migrate case for now and focus on what someone could possibly do to even backup/restore even a single PVC.  this large set of PVCs is the default configuration as you say, and it leaves very few options for a reasonable backup/restore method.  how would i even approach doing a single one of these?

**Comment by manics** [comment]
@agnewp That's a really good question that would be well suited to our [community forum](https://discourse.jupyter.org/) where more people hang out. If you start a thread there you'll be able to help or get the input of many other K8s admins across multiple systems. Depending on the conclusions from that thread we could either copy some of the suggestions to this guide, or perhaps just link out to it.

**Comment by agnewp** [comment]
so is the answer that essentially if i copy all the file inside that volume, that is a 'backup' and if i paste them back, that is a restore? i don't need k8s specifics and its clear jupyter doesn't have a mechanism for doing this, however again what things should be disabled, is 100% of the application state on this user specific volume?

**Comment by MridulS** [comment]
@agnewp one thing to keep in mind while doing the backup is that the PVC name generated could change, so while backing up user data, you would also need to backup the PVC for users. On redeploying jupyterhub you may end up with new PV names in the PVC, so while copying back data to the new PVs you would need to match the old PVC with the new PVC.

We use NFS-provisioner for user storage on our cluster and we run a backup CRON job daily to back up user data and the PVC, if its helps you can have a look at https://github.com/gesiscss/orc/blob/master/storage/backup/docker/backup.py 
We have some internal documentation for this and I was planning on moving it to a public facing site.

**Comment by mnp** [comment]
A minor footnote to this section would be how to resize a user's PVC. Something along the lines of

```shell
kubectl get pvc | grep claim-
kubectl edit pvc/claim-user1
```

Edit the value under `spec.resources.storage`. The user's lab needs to be stopped and restarted for it to take effect.

**Comment by milesgranger** [comment]
Just came across this; I think it'd be difficult to make even a general guidance of backing up user data, doesn't it almost wholly depend on what storage class is being used? For us w/ longhorn there's cron jobs to backup PVCs to another (on-prem) S3 location - using managed k8s on GCP, AWS, etc would probably have other solutions.  If it's helpful, I could make a PR to add a small snippet on this just to have this issue closed - that it's basically up to the platform engineer's discretion accounting for their current situation.

**Comment by manics** [comment]
Given that it's 4 years I think saying it's the admin's responsibility to back up volumes is reasonable. Z2JH is aimed at people with limited K8s knowledge which is why it would be nice to have instructions, but as you say it's very much dependent on the storage implementation. If someone knows of a generic solution in future we can always update the docs again!

**Comment by TiemenSch** [comment]
I still think it would be nice to offer a user backup regardless of the volume/storage class implementation of the current hub.

Backing up user volumes from the K8s side provides a solution to do user data rollback with the deployment in place, but doesn't quite give you a "backup and restore" functionality to (fairly) quickly replicate your deployment on a different infrastructure provider for example.

I agree that there is a multitude of routes that could be taken, but at the moment I believe none of them are described. 
Having one example route shown could go a long way.

**Comment by mananpreetsingh** [comment]
Yeah I agree, we just had to go through this use case of migrate JH instance, I had to ask users to  download all their content and upload it in new JH. I have developed a small script for users to do backup and restore. But did not see any native way to do so. I will write a blog article how we do it manually for now. It is manual but still easy just execute one notebook and download files.

**Comment by mnp** [comment]
This can be done en masse at the PV/PVC level too, without involving the user. Two possible approaches:
1. Valero or similar generic cluster backup tool will back up the disks
2. You could write a script to log out a user, launch a work pod like busybox that mounts that disk, copies the files somewhere, and then kills the pod.
3. You can also shell into a user's running lab and copy their home dir (/home/jovyan) somewhere.
